components/eam/src/chemistry/modal_aero/mo_spitfire_transport.F90

```diff
@@ -0,0 +1,395 @@
+
+!-----------------------------------------------------------------
+! Assumed grid staggering:
+!
+! Input:
+!
+!     xw1.......xw2.......xw3.......xw4.......xw5.......xw6
+!    vel1......vel2......vel3......vel4......vel5......vel6
+!     ....phi1......phi2.......phi3.....phi4.......phi5....
+!
+! Work arrays:
+!
+!    psi1......psi2......psi3......psi4......psi5......psi6
+!
+! Output:
+!
+!   flux1.....flux2.....flux3.....flux4.....flux5.....flux6
+!-----------------------------------------------------------------
+
+    implicit none
+
+    integer, intent(in) :: ncol     ! number of colums to process
+
+    real(r8),intent(in) ::  xw(pcols,pverp)     ! coordinate variable, values at layer interfaces. In EAM this is pint.
+    real(r8),intent(in) :: phi(pcols,pverp-1)   ! grid cell mean tracer mixing ratio
+    real(r8),intent(in) :: vel(pcols,pverp)     ! velocity in the xw coordinate. In EAM this is grav * rho * v
+                                                ! where v is velocity in the height (z) coordinate 
+    real(r8),intent(in) :: deltat
+
+    real(r8),intent(out) :: flux(pcols,pverp)
+
+    real(r8) psi(pcols,pverp)    ! integral of phi along the xw coordinate
+    real(r8) xxk(pcols,pver)     ! departure point in the xw coordinate
+    real(r8) psistar(pcols)      ! integral of phi at departure point
+    real(r8) fdot(pcols,pverp)   ! derivative of interpolating polynomial
+
+    integer :: kk  ! vertical layer index
+
+    ! Set fluxes at boundaries to zero
+
+    flux(:ncol,1) = 0
+    flux(:ncol,pverp) = 0._r8
+
+    ! Get the vertical integral of phi. 
+    ! See Rasch and Lawrence (1998), Eq (3) but note we are using a pressure coordinate here.
+
+    psi(:ncol,1) = 0._r8
+    do kk = 2,pverp
+       psi(:ncol,kk) = phi(:ncol,kk-1)*( xw(:ncol,kk)-xw(:ncol,kk-1) ) + psi(:ncol,kk-1)
+    end do
+
+    ! Calculate the derivatives for the interpolating polynomial
+
+    call cfdotmc_pro (ncol, xw, psi, &! in
+                      fdot           )! out
+
+    ! Calculate fluxes at interior interfaces
+
+    do kk = 2,pver
+
+       ! Find departure point. Rasch and Lawrence (1998), Eq (4)
+
+       xxk(:ncol,kk) = xw(:ncol,kk)-vel(:ncol,kk)*deltat   
+
+       ! Calculate the integral, psistar, at the departure point xxk.
+
+       call cfint2(ncol, xw, psi, fdot, xxk(:,kk), &! in
+                   psistar                         )! out
+
+       ! Calculate the flux at interface kk. Rasch and Lawrence (1998), Eq (5)
+
+       flux(:ncol,kk) = psi(:ncol,kk)-psistar(:ncol)
+    end do
+
+  end subroutine getflx
+
+
+
+!##############################################################################
+! Given a coordinate xw, an interpolating polynomial ff and its derivative fdot,
+! calculate the value of the polynomial (psistar) at xin. 
+!##############################################################################
+
+  subroutine cfint2 (ncol, xw, ff, fdot, xin, psistar)
```

Comment:
I think there might be room for further refactor it. Can make some functions for the calculations. Also, many of the variable names do not make much senses. It would be great if we could have some meaningful names.